### PR TITLE
Allow custom request options to be specified

### PR DESCRIPTION
**Pull request by rhodgkins**
#### Why?

So a custom HTTP agent can be used (want to allow `forever` to be used for [request](https://github.com/request/request/tree/v2.83.1#requestoptions-callback)).

#### How?

Matched the `.userPromises()` and `.useBaseUrl()` pattern and added `.useRequestOpts()`.

I've also ensured that the default options (and headers) passed into the request method are maintained.

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -132,11 +138,24 @@ export default class Client {
         'User-Agent': 'intercom-node-client/2.0.0'
       }
     };
+    // Append methods arguments
+    args = Object.assign({}, defaultArgs, args);
+    // Arguements to finally use for request
+    const requestArgs = Object.assign({}, this.requestOpts);
+    // Don't just use Object.assign(requestArgs, args) here because we need to handle special cases like .headers
```
Comment (mattdonnelly) [inline lib/client.js]: It sounds like what we're aiming for here is a deep merge rather than a shallow merge like Object.assign does. I think it'd be good to move this logic out of this function into a generic deep merge function or we could add the `merge` function from [lodash](https://lodash.com/docs/4.17.11#merge). What do you think?

File: test/request-opts.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import assert from 'assert';
+import { Client } from '../lib';
+import nock from 'nock';
+import sinon from 'sinon';
+
+describe('request-opts', () => {
```
Comment (mattdonnelly) [inline test/request-opts.js]: Would be good to swap the arrows functions out for regular functions. Using arrow functions with Mocha is discouraged: https://mochajs.org/#arrow-functions

File: yarn.lock [hunk 3]
```diff
@@ -2,34 +2,59 @@
 # yarn lockfile v1
 
 
-"@sinonjs/formatio@^2.0.0":
-  version "2.0.0"
-  resolved "http://registry.npmjs.org/@sinonjs/formatio/-/formatio-2.0.0.tgz#84db7e9eb5531df18a8c5e0bfb6e449e55e654b2"
+"@sinonjs/commons@^1.0.2":
```
Comment (mattdonnelly) [inline yarn.lock]: What are these changes needed for?

File: lib/client.js [hunk 4]
```diff
@@ -54,6 +55,10 @@ export default class Client {
     this.baseUrl = baseUrl;
     return this;
   }
+  useRequestOpts(opts) {
+    this.requestOpts = Object.assign(this.requestOpts, opts);
```
Comment (mattdonnelly) [inline lib/client.js]: This is nice 👍

File: yarn.lock (hunk 3 above)
Comment (rhodgkins) [reply yarn.lock]: If you just run a plain `yarn` (or `yarn install`), it adds these in. They're dev dependencies anyway so I guess it doesn't matter too much?

File: test/request-opts.js (hunk 2 above)
Comment (rhodgkins) [reply test/request-opts.js]: I totally agree, but I was just trying to keep it consistent with the rest of the test suite - most people prefer that! Shall I change or?

File: lib/client.js (hunk 1 above)
Comment (rhodgkins) [reply lib/client.js]: Also we could just go with:

```
args = Object.assign({}, this.requestOpts, defaultArgs, args);
args.headers = Object.assign({}, this.requestOpts.headers, args.headers);
```

As its only headers that are specified in the defaultArgs anyway?

File: test/request-opts.js (hunk 2 above)
Comment (mattdonnelly) [reply test/request-opts.js]: I think it'd be good to change it. We shouldn't be doing this elsewhere in the codebase

File: yarn.lock (hunk 3 above)
Comment (mattdonnelly) [reply yarn.lock]: I think we should exclude it from this PR if we don't need it then. Might end up being a breaking change depending on people's yarn/node versions

File: test/request-opts.js (hunk 2 above)
Comment (mattdonnelly) [reply test/request-opts.js]: Nice! Thanks. Yep, the other tests are definitely for a separate PR

File: README.md [hunk 5]
```diff
@@ -80,6 +80,27 @@ client.users.create({ email: 'foo@bar.com' }).then(function (r) {
...
+```
+
+Note that certain request options (such as `json`, and certain `headers` cannot be overriden).
+
+To change the `baseUrl` for the client library use:
+
+```node
+var client = new Intercom.Client({ token: 'my_token' });
+client.useBaseUrl('http://local.test-server.com');
+```
```
Comment (rhodgkins) [inline README.md]: @mattdonnelly after updating the README to include the `.useBaseUrl()` method, would it be worth deprecating that (with the functionality still working but with a warning) in place of using `.useRequestOpts({ baseUrl: 'blah' })`?

Comment (mattdonnelly) [reply README.md]: Good idea, I think we can remove it from the README, but we should keep the `useBaseUrl` method until we decide to make a major version release

File: lib/client.js (hunk 1 above)
Comment (mattdonnelly) [reply lib/client.js]: Looks like `lodash.merge` is only 64K so I think it should be fine to add. This is a server side SDK so we don't need to worry too much about adding dependencies.

Think it'd be good to go with the deep merge so we can get any additional object options for free. I think we should give precedence to the args passed in though. We don't want to allow things like the `json` key or URL to changeable

File: README.md (hunk 5 above)
Comment (rhodgkins) [reply README.md]: Done!

File: lib/client.js [hunk 6]
```diff
@@ -44,14 +46,19 @@ export default class Client {
     this.conversations = new Conversation(this);
     this.notes = new Note(this);
     this.promises = false;
-    this.baseUrl = 'https://api.intercom.io';
+    this.requestOpts = {
+      baseUrl: 'https://api.intercom.io'
+    };
+
+    // Keep this API around but mark it deprecated
```
Comment (mattdonnelly) [inline lib/client.js]: I think this is implied by using the deprecate function

**Comment by rhodgkins** [comment]
@mattdonnelly I forgot to delete an old test file when I deprecated that method. Hang on!

**Comment by rhodgkins** [comment]
@mattdonnelly any ETA on a release only you can't use this package via another repo (i.e. `"intercom-client": "bookcreator/intercom-node#request-opts"`) due to the way its setup :(

**Comment by choran** [comment]
@rhodgkins Thanks so much for the work here.
I will publish a new version on Monday 👍 
thanks again

**Comment by rhodgkins** [comment]
@choran any chance of that release please?
